Adds PCH API Calls for DNS Management

nils-wisiol opened the pull request:
Prepares for deployment via PCH.

Open questions:

- [x] How do we proceed when PCH API calls fail? (See code comments in PDNS change tracker.)
- [x] Can we contact PCH before we AXFR to nsmaster?
- [x] PCH calls are optional by virtue of env settings. Do we need tests covering the case where PCH calls are skipped?

@peterthomassen happy to hear your comments on these and everything else. I strongly recommend to view the change set commit by commit.

nils-wisiol commented:
As discussed offline:

> How do we proceed when PCH API calls fail? (See code comments in PDNS change tracker.)
DB transaction should be rolled back to provide consistency across dbapi and PCH.

> PCH calls are optional by virtue of env settings. Do we need tests covering the case where PCH calls are skipped?
Tests are not required for the non-PCH case. Make PCH requests conditional on the DEBUG flag from settings.

peterthomassen commented on api/desecapi/metrics.py:
```diff
@@ -74,6 +74,19 @@ def set_histogram(name, *args, **kwargs):
 set_counter("desecapi_pdns_keys_fetched", "number of times pdns keys were fetched")
 
+# pch.py metrics
+set_counter(
+    "desecapi_pch_request_success",
+    "number of times PCH request was successful",
+    ["method", "status"],
+)
+set_counter(
+    "desecapi_pch_request_failure",
+    "number of times PCH request failed",
+    ["method", "path", "status"],
+)
```
I think it's ok to keep these separate, so no need to do any work here (just summarizing from offline discussion)

peterthomassen commented on api/desecapi/pdns_change_tracker.py:
```diff
@@ -5,7 +5,7 @@
-from desecapi import metrics
+from desecapi import metrics, pch
 from desecapi.models import RRset, RR, Domain
 from desecapi.pdns import (
```
`from desecapi import pdns`

nils-wisiol replied:
done & refactored to avoid usage of private API and for better separation of concerns

peterthomassen commented on api/desecapi/tests/base.py:
```diff
@@ -609,6 +611,62 @@ def request_pdns_update_catalog(cls):
+                    json.dumps(
+                        {
+                            "status": True,
+                            "message": "some message",
```
The message is `"Zone(s) deleted`"` (probably doesn't matter)

nils-wisiol replied:
adjusted

nils-wisiol commented on api/desecapi/pch.py:
```diff
@@ -0,0 +1,66 @@
+    return _pch_request("post", path=path, data=data, **kwargs)
+
+
+def _patch(path, data, **kwargs):
+    return _pch_request("patch", path=path, data=data, **kwargs)
+
+
+def _get(path, **kwargs):
+    return _pch_request("get", path=path, **kwargs)
+
+
+def _put(path, **kwargs):
+    return _pch_request("put", path=path, **kwargs)
```
done

nils-wisiol commented on api/desecapi/pdns_change_tracker.py:
```diff
@@ -304,6 +316,10 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+                if settings.PCH_API:
+                    # Note that we do not have tests covering the case that this is skipped
+                    change.pch_do()  # TODO discuss if we want PCH failures to fail the transaction;
+                    #      discuss if PCH calls before AXFR are okay
```
Waiting for feedback from PCH

nils-wisiol commented:
I've addressed the code comments and cleaned up our `pdns.py` / `pdns_change_tracker.py` interaction a little.

Changes also include rebasing on #675 and adding the env variables to `docker-compose.yml`. (This will fix the GH API tests, I hope.)

Currently the only open item is the discussion of AXFR order, where we wait for confirmation from PCH. If they confirm, this can be merged.

peterthomassen commented on api/desecapi/tests/test_domains.py:
```diff
@@ -270,9 +270,7 @@ def test_list_domains_owns_qname(self):
-        with self.assertRequests(
-            self.requests_desec_domain_deletion(self.my_domain)
-        ):
+        with self.assertRequests(self.requests_desec_domain_deletion(self.my_domain)):
```
This commit should be squashed into 5b177d628ffcba082d91c04ef138e90e6402a3aa

nils-wisiol replied:
done

peterthomassen commented on api/desecapi/pdns_change_tracker.py:
```diff
@@ -7,17 +7,7 @@
-    construct_catalog_rrset,
-)
-
+from desecapi import pdns
```
should go in line 8 along with other `desecapi` imports

peterthomassen commented on api/desecapi/pdns.py:
```diff
@@ -191,6 +192,86 @@ def get_rrset_datas(domain):
+
+
+def create_zone_lord(name):
+    name = name.rstrip(".") + "."
```
Consider computing this in `CreateDomain.pdns_do()` (which is also pdns-specific) and passing it here and to `create_zone_master`

nils-wisiol replied:
AFAIK, the following HTTP request cannot be used without the final dot on the name. Moving this away would make the API exposed by pdns.py more error-prone.

peterthomassen replied:
sure, fine with me

peterthomassen commented on api/desecapi/pdns.py:
```diff
@@ -191,6 +192,86 @@ def get_rrset_datas(domain):
+        {
+            "name": name,
+            "kind": "SLAVE",
+            "masters": [socket.gethostbyname("nslord")],
```
Consider doing the lookup once at startup, then store in settings

nils-wisiol replied:
done

peterthomassen commented on api/desecapi/pdns_change_tracker.py:
```diff
@@ -72,67 +71,15 @@ def api_do(self):
-                pdns.NSMASTER,
-                "/zones?rrsets=false",
-                {
-                    "name": self.domain_name_normalized,
```
The `domain_name_normalized` property is now obsolete and can be removed

nils-wisiol replied:
removed

nils-wisiol replied:
removed

peterthomassen commented on api/desecapi/pdns_change_tracker.py:
```diff
@@ -217,7 +164,7 @@ def pdns_do(self):
             }
 
             if data["rrsets"]:
-                pdns._pdns_patch(pdns.NSLORD, "/zones/" + self.domain_pdns_id, data)
```
The `domain_pdns_id` property is now obsolete and can be removed

peterthomassen reviewed (changes requested):
Cool! Only very minor stuff, otherwise waiting for PCH.

nils-wisiol commented:
Fixed all code comments

peterthomassen commented:
I believe you accidentally committed a bunch of .ipynb files in api/ and api/.ipynb_checkpoints/.

nils-wisiol commented:
woops, sorry about that. I fixed it now.
